### PR TITLE
[PatternLang] Add Syntatic Sugar to the C++ pattern API and support DataType Attribute Matching

### PR DESCRIPTION
**Pull request by mbrookhart**
cc @icemelon9

### CODE REVIEW

File: include/tvm/relay/dataflow_pattern.h [hunk 1]
```diff
@@ -1,5 +1,4 @@
-/*
- * Licensed to the Apache Software Foundation (ASF) under one
+/* * Licensed to the Apache Software Foundation (ASF) under one
```
Comment (icemelon9) [inline include/tvm/relay/dataflow_pattern.h:1]: revert this change

File: tests/cpp/dataflow_pattern_test.cc [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <gtest/gtest.h>
+#include <tvm/relay/dataflow_pattern.h>
+#include <tvm/tir/analysis.h>
+
```
Comment (icemelon9) [inline tests/cpp/dataflow_pattern_test.cc:23]: Add a unit test for call?

File: include/tvm/relay/dataflow_pattern.h [hunk 3]
```diff
@@ -46,6 +48,29 @@ class DFPatternNode : public Object {
  */
 class DFPattern : public ObjectRef {
  public:
+  /*! \brief Syntatic Sugar for creating a CallPattern */
+  DFPattern operator()(const std::vector<DFPattern>& args);
+  /*! \brief Syntatic Sugar for creating a CallPattern with an "add" op */
```
Comment (icemelon9) [inline include/tvm/relay/dataflow_pattern.h:53]: I feel that the syntactic sugar "+-*/" is not necessary since we have the "IsOp" one, and it may cause some confusion whether you're adding two patterns together or use the "add" op.

**Review by icemelon9 — COMMENTED** [review]
overall lgtm.

File: tests/cpp/dataflow_pattern_test.cc (hunk 2 above)
Comment (mbrookhart) [reply tests/cpp/dataflow_pattern_test.cc:23]: I didn't add one because all of the operator +-*/ ops go through it.

File: include/tvm/relay/dataflow_pattern.h (hunk 3 above)
Comment (mbrookhart) [reply include/tvm/relay/dataflow_pattern.h:53]: Hmm. I did the +-*/ overloads in python because Relay's IR does the same thing at the python level, and I wanted to keep operation on the pattern API in line with Relay's API. I continued that here, trying to keep the C++ pattern API in line with the Python Pattern API. So...for consistency I'd argue to leave it as it is.

Comment (icemelon9) [reply include/tvm/relay/dataflow_pattern.h:54]: I see. sure.

**Review by icemelon9 — APPROVED** [review]
LGTM. Could you rebase to master?

**Comment by icemelon9** [comment]
Thanks @mbrookhart
